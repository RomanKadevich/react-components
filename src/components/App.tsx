import { useCallback, useEffect, useState } from "react";
import { IAnimal, IAnimals } from "../types/types";
import { List } from "./List";
import Pagination from "./Pagination";
import { useNavigate, useParams, useSearchParams } from "react-router-dom";

interface IAppState {
  data: IAnimal[];
  error: Error | null;
  isLoading: boolean;
  pageNumber: number;
}
const API_BASE_URL = "https://stapi.co/api/v1/rest/animal/search/";

const PAGE_SIZE = 12;

export const App = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const name = searchParams.get("name");
  const details = searchParams.get("details");
  const { page } = useParams();
  const navigate = useNavigate();
  const initialState: IAppState = {
    data: [],
    error: null,
    isLoading: false,
    pageNumber: 0,
  };
  const [state, setState] = useState(initialState);
  const getData = useCallback(
    async (search?: string | null) => {
      let response: Response | undefined = undefined;
      if (!search) {
        response = await fetch(
          `${API_BASE_URL}?&pageNumber=${
            page ? +page - 1 : 1
          }&pageSize=${PAGE_SIZE}`,
        );
      } else {
        response = await fetch(
          `${API_BASE_URL}?&pageNumber=${
            page ? +page - 1 : 1
          }&name=${search}&pageSize=${PAGE_SIZE}`,
          {
            method: "POST",
          },
        );
      }
      if (!response.ok) {
        throw new Error("Network response was not ok");
      }

      const animals: IAnimals = await response.json();
      return animals;
    },
    [page],
  );
  useEffect(() => {
    const loadData = async () => {
      try {
        setState((prevState) => ({ ...prevState, isLoading: true }));
        const lastQueryData: string | null = localStorage.getItem("lastQuery");
        if (lastQueryData) {
          const animals = await getData(lastQueryData);

          setState((prevState) => ({
            ...prevState,
            data: animals.animals,
            error: null,
            isLoading: false,
            pageNumber: animals.page.totalPages,
          }));
        } else {
          searchParams.set("name", "");
          setSearchParams(searchParams);
          const animals: IAnimals = await getData(name);
          setState((prevState) => ({
            ...prevState,
            data: animals.animals,
            error: null,
            isLoading: false,
            pageNumber: animals.page.totalPages,
          }));
        }
      } catch (error) {
        if (error instanceof Error) {
          setState((prev) => ({ ...prev, error: error as Error, data: [] }));
        }
      }
    };
    loadData();
  }, [getData, navigate, name, page, setSearchParams, searchParams]);

  return (
    <div
      onClick={() => {
        if (details) setSearchParams({ name: name || "", details: "" });
      }}
      className={details ? "blur-sm" : ""}
    >
      {state.error ? (
        <div className="bg-white p-3 text-center text-red-500 font-bold">
          Error when loading data{state.error.message}
        </div>
      ) : null}
      {state.data ? (
        <>
          {state.isLoading && (
            <div className="w-full h-full  flex justify-center  font-bold text-lg relative bg-black ">
              <p className="absolute top-[20vh] text-[2rem]">Loading...</p>
            </div>
          )}
          <List animals={state.data} />
          <Pagination
            pageNumber={state.pageNumber}
            pageIndex={page ? +page : 1}
          />
        </>
      ) : null}
    </div>
  );
};

export default App;
