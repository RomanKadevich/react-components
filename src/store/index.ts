import { configureStore } from "@reduxjs/toolkit";
import searchValueSlice from "./slices/searchValueSlice";

const store = configureStore({
  reducer: {
    value: searchValueSlice,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
