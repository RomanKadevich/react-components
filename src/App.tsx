import { Component } from "react";
import Header from "./components/header";
import {
  HandlerInputType,
  HandlerSubmitType,
  IAnimal,
  IAnimals,
} from "./types/types";
import { List } from "./components/gallery";
interface IAppState {
  value: string;
  data: IAnimal[];
}
export class App extends Component {
  state: IAppState = {
    value: "",
    data: [],
  };
  componentDidMount = async () => {
    const lastSearchData: string | null = localStorage.getItem("lastSearch");
    const lastQueryData: string | null = localStorage.getItem("lastQuery");
    if (lastSearchData) {
      const lastSearch: IAnimal[] = JSON.parse(lastSearchData);
      this.setState({ data: lastSearch });
      if (lastQueryData) {
        this.setState({ value: lastQueryData });
      }
    } else {
      try {
        const response = await fetch(
          "http://stapi.co/api/v1/rest/animal/search?pageNumber=1&pageSize=12"
        );
        const animals: IAnimals = await response.json();
        this.setState({ data: animals.animals });
      } catch (err) {
        console.error(err);
      }
    }
  };
  handleInputSearch: HandlerInputType = (event) => {
    this.setState({ value: event.currentTarget.value });
    console.log(this.state.value);
  };
  handleSubmitSearch: HandlerSubmitType = async (event) => {
    event.preventDefault();
    try {
      const response = await fetch(
        `http://stapi.co/api/v1/rest/animal/search?name=${this.state.value}&pageSize=12`,
        {
          method: "POST",
        }
      );
      const animals: IAnimals = await response.json();
      console.log(animals.animals);
      this.setState({ data: animals.animals });
      localStorage.setItem("lastSearch", JSON.stringify(animals.animals));
      localStorage.setItem("lastQuery", this.state.value);
    } catch (error) {
      console.error(error);
    }
  };

  render() {
    const { value, data } = this.state;
    return (
      <>
        <Header
          handleInputSearch={this.handleInputSearch}
          handleSubmitSearch={this.handleSubmitSearch}
          value={value}
        />
        <List animals={data} />
      </>
    );
  }
}

export default App;
