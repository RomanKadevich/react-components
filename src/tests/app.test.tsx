import { afterAll, afterEach, beforeAll, describe, expect, it } from "vitest";
import { render, screen, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { server } from "./handlers";
import App from "../App";
import { Provider } from "react-redux";
import store from "../store";

describe("App", () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should check correct list item", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const list = await screen.findAllByTestId("card-item");
    expect(list).toHaveLength(12);
  });
  it("should show no cards if no data", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const searchBtn = await screen.findByTestId("search");
    const input = await screen.findByTestId("input");
    const badQuery = "fwefqweqwef";
    act(() => {
      fireEvent.change(input, { target: { value: badQuery } });
      fireEvent.click(searchBtn);
    });
    const noCardsNote = await screen.findByText("No cards");
    expect(noCardsNote).toBeInTheDocument();
  });
  it("should render the relevant card data", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const searchBtn = await screen.findByTestId("search");
    const input = await screen.findByTestId("input");
    const Query = "Albatross";
    act(() => {
      fireEvent.change(input, { target: { value: Query } });
      fireEvent.click(searchBtn);
    });
    const title = await screen.findByText("Abalone");

    expect(title).toBeTruthy();
    const firstCard = 1;
    const list = await screen.findAllByTestId("card-text-list");
    const item = list[firstCard];
    expect(item.textContent).toContain("Canine");
    expect(item.textContent).toContain("Earth Animal");
    expect(item.textContent).toContain("Feline");
    expect(item.textContent).toContain("Earth Insect");
  });
  it("should render detail component on click", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const card = await screen.findAllByTestId("card-item");
    const firstCard = 0;
    act(() => {
      fireEvent.click(card[firstCard]);
    });
    const detailCard = await screen.findByTestId("details");

    expect(detailCard).toBeInTheDocument();
  });

  it("should render loading during fetch details", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const card = await screen.findAllByTestId("card-item");
    const firstCard = 0;
    act(() => {
      fireEvent.click(card[firstCard]);
    });

    expect(card).toBeTruthy();
  });
  it("should render correct data in the details", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const card = await screen.findAllByTestId("card-item");
    const firstCard = 1;
    act(() => {
      fireEvent.click(card[firstCard]);
    });
    const detailCard = await screen.findByTestId("details");
    expect(detailCard.textContent).toContain("Canine");
    expect(detailCard.textContent).toContain("Earth Animal");
    expect(detailCard.textContent).toContain("Feline");
    expect(detailCard.textContent).toContain("Earth Insect");
  });

  it("should hide the  details by clicking on the area outside", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const card = await screen.findAllByTestId("card-item");
    const firstCard = 1;

    act(() => {
      fireEvent.click(card[firstCard]);
    });

    const outsideAreaDetailCard = await screen.findByTestId("outside-details");
    act(() => {
      fireEvent.click(outsideAreaDetailCard);
    });
    const closeDetailCardWrapper = await screen.findByTestId("details-wrapper");
    expect(closeDetailCardWrapper).toHaveClass("hidden");
    act(() => {
      fireEvent.click(card[firstCard]);
    });
  });

  it("should update URL query parameter when page changes", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    await screen.findAllByTestId("card-item");
    const nextBtn = await screen.findByTestId("next");
    act(() => {
      fireEvent.click(nextBtn);
    });
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    expect(window.location.pathname).toBe("/1");
  });
  it("should Ð¼erify that clicking the Search button saves the entered value to the local storage", async () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const searchBtn = await screen.findByTestId("search");
    const input = await screen.findByTestId("input");
    const Query = "test";
    act(() => {
      fireEvent.change(input, { target: { value: Query } });
      fireEvent.click(searchBtn);
    });
    const storedData = localStorage.getItem("lastQuery");
    expect(storedData).toEqual(Query);
  });
});
