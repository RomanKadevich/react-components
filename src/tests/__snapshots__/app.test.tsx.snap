// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should create List 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="container mx-auto flex justify-center flex-wrap gap-3 p-[20px]"
      >
        <button
          class="cursor-pointer hover:opacity-60"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.5 12h-15m0 0l6.75 6.75M4.5 12l6.75-6.75"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        <li
          class=" bg-violet-500 cursor-pointer p-1.5 rounded"
          data-testid="pagBtn-0"
        >
          1
        </li>
        <button
          class="cursor-pointer hover:opacity-60"
          data-testid="next"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.5 12h15m0 0l-6.75-6.75M19.5 12l-6.75 6.75"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="container mx-auto flex justify-center flex-wrap gap-3 p-[20px]"
    >
      <button
        class="cursor-pointer hover:opacity-60"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.5 12h-15m0 0l6.75 6.75M4.5 12l6.75-6.75"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <li
        class=" bg-violet-500 cursor-pointer p-1.5 rounded"
        data-testid="pagBtn-0"
      >
        1
      </li>
      <button
        class="cursor-pointer hover:opacity-60"
        data-testid="next"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.5 12h15m0 0l-6.75-6.75M19.5 12l-6.75 6.75"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App > should render Pagination 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="container mx-auto flex justify-center flex-wrap gap-3 p-[20px]"
      >
        <button
          class="cursor-pointer hover:opacity-60"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.5 12h-15m0 0l6.75 6.75M4.5 12l6.75-6.75"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        <li
          class=" bg-violet-500 cursor-pointer p-1.5 rounded"
          data-testid="pagBtn-0"
        >
          1
        </li>
        <button
          class="cursor-pointer hover:opacity-60"
          data-testid="next"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.5 12h15m0 0l-6.75-6.75M19.5 12l-6.75 6.75"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="container mx-auto flex justify-center flex-wrap gap-3 p-[20px]"
    >
      <button
        class="cursor-pointer hover:opacity-60"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.5 12h-15m0 0l6.75 6.75M4.5 12l6.75-6.75"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <li
        class=" bg-violet-500 cursor-pointer p-1.5 rounded"
        data-testid="pagBtn-0"
      >
        1
      </li>
      <button
        class="cursor-pointer hover:opacity-60"
        data-testid="next"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.5 12h15m0 0l-6.75-6.75M19.5 12l-6.75 6.75"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
